// react-router-redux/ConnectedRouter.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRouter = require('react-router');

var _reducer = require('./reducer');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ConnectedRouter = function (_Component) {
  _inherits(ConnectedRouter, _Component);

  function ConnectedRouter() {
    _classCallCheck(this, ConnectedRouter);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  ConnectedRouter.prototype.render = function render() {
    var _props = this.props,
        propsStore = _props.store,
        history = _props.history,
        children = _props.children,
        props = _objectWithoutProperties(_props, ['store', 'history', 'children']);

    var store = propsStore || this.context.store;

    return _react2.default.createElement(
      _reactRouter.Router,
      _extends({}, props, { history: history }),
      _react2.default.createElement(_reactRouter.Route, { render: function render(_ref) {
          var location = _ref.location;

		  // custom replaced
          /*store.dispatch({
            type: _reducer.LOCATION_CHANGE,
            payload: location
          });*/
			setTimeout(()=> {
				store.dispatch({
					type: _reducer.LOCATION_CHANGE,
					payload: location,
				});
			});

          return children;
        } })
    );
  };

  return ConnectedRouter;
}(_react.Component);

ConnectedRouter.propTypes = {
  store: _react.PropTypes.object,
  history: _react.PropTypes.object,
  children: _react.PropTypes.node
};
ConnectedRouter.contextTypes = {
  store: _react.PropTypes.object
};
exports.default = ConnectedRouter;
